common/transport/lwm2m/src/main/java/org/thingsboard/server/transport/lwm2m/server/client/LwM2mClient.java

```diff
@@ -241,36 +243,71 @@ public ObjectModel getObjectModel(String pathIdVer, LwM2mModelProvider modelProv
                 .getObjectModel(pathIds.getObjectId()) : null;
     }
 
-    public String objectToString(LwM2mObject lwM2mObject) {
+    public Optional<String> contentToString(Object content) {
+        String value = null;
+        LwM2mResource resource = null;
+        String key = null;
+        if (content instanceof HashMap && ((HashMap) content).size() == 1) {
+            key = ((HashMap) content).keySet().toArray()[0].toString();
+            if (((HashMap) content).values().toArray()[0] != null) {
+                if (((HashMap) content).values().toArray()[0] instanceof LwM2mResource) {
+                    resource = (LwM2mResource) ((HashMap) content).values().toArray()[0];
+                }
+            }
+        } else if (content instanceof LwM2mResource) {
+            resource = (LwM2mResource) content;
+        }
+        if (resource != null && resource.getType() == OPAQUE) {
+            value = this.resourceToString(resource, key);
+        }
+        value = value == null ? content.toString() : value;
+        return Optional.of(String.format("%s", value));
+    }
+
+    public String resourceToString(LwM2mResource resource, String key) {
+        String value = null;
         StringBuilder builder = new StringBuilder();
-        builder.append("LwM2mObject [id=").append(lwM2mObject.getId()).append(", instances={");
-        lwM2mObject.getInstances().forEach((instId, inst) -> {
-            builder.append(instId).append("=").append(this.instanceToString(inst)).append(", ");
-        });
-        int startInd = builder.lastIndexOf(", ");
-        if (startInd > 0) {
-            builder.delete(startInd, startInd + 2);
+        if (resource instanceof LwM2mSingleResource && ((byte[]) resource.getValue()).length > 0) {
```

Comment:
(byte[]) resource.getValue() may cause ClassCastException. value may be string or something else.